src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,308 @@
+            DateTime startTime,
+            IDateTimeProvider dateTimeProvider,
+            TransactionOptions transactionOptions,
+            CancellationToken cancellationToken)
+        {
+            TResult result;
+
+            while (true)
+            {
+                clientSession.StartTransaction(transactionOptions);
+
+                var transactionStepResult = await WithTransactionCallbackProcessingAsync(clientSession, callbackAsync, startTime, dateTimeProvider, cancellationToken).ConfigureAwait(false);
+                if (transactionStepResult.TransactionShouldBeRestarted)
+                {
+                    continue;
+                }
+
+                result = transactionStepResult.Result;
+
+                if (!clientSession.IsInTransaction)
+                {
+                    return result; // Assume callback intentionally ended the transaction.
+                }
+
+                var transactionHasBeenConfirmed = await WithTransactionCommitProcessingWithRetryAsync(clientSession, startTime, dateTimeProvider, cancellationToken).ConfigureAwait(false);
+                if (!transactionHasBeenConfirmed)
+                {
+                    // the transaction has been confirmed.
+                    break;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            return result;
+        }
+
+        private static T GetElementOrDefault<T>(BsonDocument document, string name) where T : BsonValue
+        {
+            return document.TryGetValue(name, out var value) ? (T) value : null;
+        }
+
+        private static bool HasTimedOut(DateTime startTime, DateTime currentTime)
+        {
+            return (currentTime - startTime) >= __withTransactionTimeout;
+        }
+
+        private static bool IsWriteConcernTimeoutError(MongoException exception)
+        {
+            if (exception is MongoWriteConcernException writeConcernException && TryParseWriteConcernError(writeConcernException.WriteConcernResult.Response, out bool isWTimeout))
+            {
+                return isWTimeout;
+            }
+
+            var writeConcernError =
+                (exception as MongoWriteException)?.WriteConcernError ??
+                (exception as MongoBulkWriteException)?.WriteConcernError;
+
+            return TryParseWriteConcernError(writeConcernError?.Details, out isWTimeout) && isWTimeout;
+        }
+
+        private static bool TryParseWriteConcernError(BsonDocument errorDetails, out bool isWTimeout)
+        {
+            const int timeoutCode = 64;
+
+            isWTimeout = false;
+            if (errorDetails == null) return false;
+
+            var writeConcernBson = GetElementOrDefault<BsonDocument>(errorDetails, "writeConcernError");
+            if (writeConcernBson != null)
+            {
+                var code = GetElementOrDefault<BsonValue>(writeConcernBson, "code")?.AsInt32;
```

Comment:
It's safer to use ToInt32() because the server isn't always very strict about whether it returns 32-bit or 64-bit integers (or even a double).

Reply:
Done